### PR TITLE
[refactor](resource) use resource to create external catalog

### PR DESCRIPTION
**Pull request by AshinGau**
# Proposed changes
## Problem summary

Use resource to create external catalog.  The previous syntax can still be used, but it has been deleted from the document and is not recommended for users.
**Create Resource**
```
-- HMS
mysql> create resource hms_resource properties(
    -> "type"="hms",
    -> 'hive.metastore.uris' = 'thrift://172.21.0.44:7004',
    -> 'dfs.nameservices'='HANN',
    -> 'dfs.ha.namenodes.HANN'='nn1,nn2',
    -> 'dfs.namenode.rpc-address.HANN.nn1'='172.21.0.32:4007',
    -> 'dfs.namenode.rpc-address.HANN.nn2'='172.21.0.44:4007',
    -> 'dfs.client.failover.proxy.provider.HANN'='org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider'
    -> );

-- MYSQL
mysql> create resource mysql_resource properties (
    -> "type"="jdbc",
    -> "user"="root",
    -> "password"="123456",
    -> "jdbc_url" = "jdbc:mysql://127.0.0.1:3316/doris_test?useSSL=false",
    -> "driver_url" = "https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/jdbc_driver/mysql-connector-java-8.0.25.jar",
    -> "driver_class" = "com.mysql.cj.jdbc.Driver");

-- ES
mysql> create resource es_resource properties (
    -> "type"="es",
    -> "hosts"="http://127.0.0.1:29200",
    -> "nodes_discovery"="false",
    -> "enable_keyword_sniff"="true");
```
**Create External Catalog**
```
mysql> create catalog hive with resource hms_resource;
Query OK, 0 rows affected (0.01 sec)

mysql> show catalog hive;
+------------------------------------------------+---------------------------------------------------------------------------+
| Key                                            | Value                                                                     |
+------------------------------------------------+---------------------------------------------------------------------------+
| resource                                       | hms_resource                                                              |
| dfs.ha.namenodes.HANN                          | nn1,nn2                                                                   |
| hive.metastore.uris                            | thrift://172.21.0.44:7004                                                 |
| type                                           | hms                                                                       |
| dfs.nameservices                               | HANN                                                                      |
| dfs.namenode.rpc-address.HANN.nn1              | 172.21.0.32:4007                                                          |
| dfs.namenode.rpc-address.HANN.nn2              | 172.21.0.44:4007                                                          |
| dfs.client.failover.proxy.provider.HANN        | org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider |
+------------------------------------------------+---------------------------------------------------------------------------+
```
**Drop Resource**
Can't drop resource when it's used by external catalog, and should drop catalog before dropping resource.
```
mysql> drop resource hms_resource;
ERROR 1105 (HY000): errCode = 2, detailMessage = Resource hms_resource is used by: hive@CATALOG
```

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.5 seconds
 load time: 425 seconds
 storage size: 17123356946 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221209150540_clickbench_pr_61173.html

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Resource.java [hunk 1]
```diff
@@ -78,6 +107,26 @@ public static Resource fromStmt(CreateResourceStmt stmt) throws DdlException {
...
+        String fullName = referenceName + REFERENCE_SPLIT + type.name();
+        if (references.remove(fullName) != null) {
+            LOG.info("Reference(type={}, name={}) is removed from resource {}, current set: {}",
+                    type, referenceName, name, references);
+            return true;
+        }
+        return false;
+    }
+
+    public synchronized boolean addReference(String referenceName, ReferenceType type) throws AnalysisException {
```
[the remark this answers is not in the document]
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Resource.java:120]: Different statements may require different parameters even when using the same resource, so some checks are added in `addReference`.

**Review by morningman — DISMISSED** [review]
LGTM

File: docs/en/docs/sql-manual/sql-reference/Data-Definition-Statements/Create/CREATE-CATALOG.md [hunk 2]
```diff
@@ -37,53 +37,58 @@ This statement is used to create an external catalog
...
-   CREATE CATALOG hive PROPERTIES (
-		"type"="hms",
-		'hive.metastore.uris' = 'thrift://172.21.0.1:7004',
-		'dfs.nameservices'='HDFS8000871',
-		'dfs.ha.namenodes.HDFS8000871'='nn1,nn2',
-		'dfs.namenode.rpc-address.HDFS8000871.nn1'='172.21.0.2:4007',
-		'dfs.namenode.rpc-address.HDFS8000871.nn2'='172.21.0.3:4007',
-		'dfs.client.failover.proxy.provider.HDFS8000871'='org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider'
+	```sql
+	CREATE RESOURCE hms_resource PROPERTIES (
```
Comment (morningman) [inline docs/en/docs/sql-manual/sql-reference/Data-Definition-Statements/Create/CREATE-CATALOG.md:54]: You need to reserve the old way to create catalog

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**